[mlir] Improve `GreedyPatternRewriteDriver` and pass documentation

matthias-springer opened the pull request:
Clarify what kind of IR modifications are allowed. Also improve the documentation of the greedy rewrite driver entry points.

Addressing comments in #76219.

matthias-springer commented on mlir/docs/PassManagement.md:
```diff
@@ -23,16 +23,21 @@ further below. All passes in MLIR derive from `OperationPass` and adhere to the
+    block.
+    *   Other threads may be operating on these operations simultaneously.
+    *   The attributes of the current operation may be modified freely.
+    *   The operands of the current operation may be modified, as long as no
```
This is a bit sus. Example:

- Pass instance 1: `currentOp->setOperand(0, currentOp->getBlock()->getArgument(0)`
- Pass instance 2: `print(currentOp->getBlock()->getArgument(0).getUses().size()`

Instance 1 modifies the current op. That is allowed. (Or is it? The use list of a parent block argument changes. Does this count as "modifying per parent op"?).

Instance 2 reads the number of uses of a parent block argument. That is allowed. But there's a race condition if both instances run in parallel.

Maybe the only modifications that we should allow are "modifying attributes"?

joker-eph replied:
You can't touch operands, this would race in the parent block.

matthias-springer replied:
Ok, that's what I thought... So the only permissible change to the operation that the pass operates on is "modifying attributes". I updates the docs accordingly.

matthias-springer committed:
[mlir] Improve `GreedyPatternRewriteDriver` and pass documentation

Clarify what kind of IR modifications are allowed. Also improve the documentation of the greedy rewrite driver entry points.

jpienaar commented:
Thanks!
